### PR TITLE
Automatically enable smell in reek_of matcher

### PR DESCRIPTION
**Pull request by mvz**
#505.

### CODE REVIEW

File: spec/reek/smells/unused_private_method_spec.rb [hunk 1]
```diff
@@ -37,19 +31,8 @@ def drive; end
...
-      expect(first_warning.smell_type).to eq(Reek::Smells::UnusedPrivateMethod.smell_type)
-      expect(first_warning.parameters[:name]).to eq(:drive)
-      expect(first_warning.lines).to eq([4])
-
-      second_warning = examiner.smells.last
-      expect(second_warning.smell_type).to eq(Reek::Smells::UnusedPrivateMethod.smell_type)
-      expect(second_warning.parameters[:name]).to eq(:start)
-      expect(second_warning.lines).to eq([3])
+      expect(source).to reek_of(:UnusedPrivateMethod, { name: :drive, lines: [4] })
+      expect(source).to reek_of(:UnusedPrivateMethod, { name: :start, lines: [3] })
```
Comment (troessner) [inline spec/reek/smells/unused_private_method_spec.rb]: Reads a lot nicer, awesome

**Comment by troessner** [comment]
Travis is red...code looks good though.

**Comment by mvz** [comment]
Yeah, I accidentally broke everything :smile:. Should be fixed now.

**Comment by mvz** [comment]
I :heart: RuboCop.

**Comment by mvz** [comment]
We may need to learn a little more about mutant. /cc @mbj

**Comment by mbj** [comment]
> We may need to learn a little more about mutant. /cc @mbj

Meet me in mutation testing Slack channel to discuss the questions?

**Comment by troessner** [comment]
>> We may need to learn a little more about mutant.

Looking at the output:

>> Coverage:        93.15%
Expected:        100.00%
The command "bundle exec mutant --include lib --require reek --use rspec --since master --jobs 4 "Reek*"" exited with 1.

and especially at the mutants like this one:

>>
-  self[detector] = (self[detector] || {}).merge(config)
+  self[detector] = (self[detector] || self).merge(config)

it's obvious that our tests are not 100% bulletproof. Not sure if there is anything we can do about besides...writing more tests.
Problem is that we have been ignoring mutant so far because having a 100% spec coverage for mutation testing is just simply exhausting 😆 (and it also takes a lot of fun away for a non-commercial project)
But then it's kind of pointless. Maybe we should switch back to whitelisting only API relevant files?

**Comment by mbj** [comment]
> Problem is that we have been ignoring mutant so far because having a 100% spec coverage for mutation testing is just simply exhausting

I'd argue it is not, since you likely fell into the most common mistake using it:

Mutation testing is NOT about adding tests, its about removing or structuring code better so the mutation cannot occur anymore. The idea is that an alive mutation only when nothing else works should be cured with adding (or changing) a test.

Mutant "always" (there are some well choosen exceptions) reduces semantics. If mutant finds a an alive reduction, for most cases you can simply use the reduction in your code and kill the mutant without adding a test.

If the reduction cannot be killed via code change (because the code would turn into invalid semantics): You miss an important test.

And adding that important test is to complex, the design is smelly and should be fixed.

Obviously I argue without detailed understanding of that specific case, but for educational experience (on both sides): What about arranging a small pairing session where we work out how the case you cited applies to the above?

**Comment by troessner** [comment]
>> I'd argue it is not, since you likely fell into the most common mistake using it:
Mutation testing is NOT about adding tests, its about removing or structuring code better so the mutation cannot occur anymore. The idea is that an alive mutation only when nothing else works should be cured with adding (or changing) a test.

Hmmm. Good point. Seems like we misunderstood this indeed. At least I did 😆 

>> Obviously I argue without detailed understanding of that specific case, but for educational experience (on both sides): What about arranging a small pairing session where we work out how the case you cited applies to the above?

Awesome idea! I'll nudge you as soon as I'll tackle the next issue in Reek, thanks!

**Pull request merged**